## Analyzing borrowers’ risk of defaulting

This project is to prepare a report for a bank’s loan division. We’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

In this report we are building a **credit scoring** of a potential customer. A **credit scoring** is used to evaluate the ability of a potential borrower to repay their loan.

### Step 1. Open the data file and have a look at the general information. 

In [1]:
#Let’s read the table from the csv file and then check: what we have in the table:

import pandas as pd
credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_scoring.info()

#We might check how this data looks like and what contecnt is in top 5 rows: 
#print(credit_scoring.head())
display(credit_scoring.head()) #very nice, thank you :)
#Let's check the length of the table - we see
print(len(credit_scoring))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

21525


### Conclusion

Data set is saved as a credit_scoring variable. Its total length is 21525 rows. The clients’ family status appears in the ‘family_status’ column and 'family_status_id'; their total income, in the ‘total_income’ column (as float); and their ages (as integers), in the ‘dob_years’ column; number of children in 'children' column, employment period in 'days_employed', educational level in 'education' column and 'education_id', client's gender in 'gender' column; source of income in 'income_type'; status of the credit repayment in 'debt' column; credit's purpose in 'purpose'. Let's see the quality of the data, for example whether there are missing values and other peculiarities.

### Step 2. Data preprocessing

### Processing missing values

In [2]:
#Let's identify and fill in missing values. 
#We sart with 'total_income' and then 'days_employed'.

print(len(credit_scoring[credit_scoring['days_employed'].isnull()]))
print(len(credit_scoring[credit_scoring['total_income'].isnull()]))
print()
#Null values of total_income and days_employed - 2174, correlated.
#Let's make sure there is correctly stored as NaN so we can operate it later on:
pd.to_numeric(credit_scoring['days_employed'], errors = 'coerce')
pd.to_numeric(credit_scoring['total_income'], errors = 'coerce')

2174
2174



0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21520    35966.698
21521    24959.969
21522    14347.610
21523    39054.888
21524    13127.587
Name: total_income, Length: 21525, dtype: float64

In [3]:
credit_scoring.info()

#We can see that while checking 'gender' there is a value XNA that worth checking.
print(credit_scoring[credit_scoring['gender']=='XNA']) #so we can delete it as it represents less than 1% of the data
credit_scoring.drop(credit_scoring[credit_scoring['gender']=='XNA'].index, axis=0, inplace=True)    
#print(credit_scoring[credit_scoring['gender']=='XNA'])

#We can see that while checking 'dob_years' there are 101 entries that have value ‘0’ for the age (‘dob_years’)
#print(credit_scoring['dob_years'].value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
       children  days_employed  dob_years     education  education_id  \
10701         0   -2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  

### Conclusion

Most cells contains 21525 rows except total_income and days_employed - 19351. These are missing values that are correlared and that we turned to NaN with 'coerce'. Next missing value   was found while checking 'gender': XNA, this row was deleted. Other columns' data looks good.




### Data type replacement

In [4]:
    
#To make sure that our data set is clean, let's see what kind of values we have in each column with value_counts() function.
#print(credit_scoring['purpose'].value_counts())

#We can see that while checking 'days_employed' there are entries with negative values.
#print(credit_scoring['days_employed'].value_counts())
#print((credit_scoring[credit_scoring['days_employed'] < 0]).count())
#There are 15905 negative values that seems to be an error of data saving, so we can turn them to positive
pd.to_numeric(credit_scoring['days_employed'], errors = 'coerce')
credit_scoring['days_employed'] = abs(credit_scoring['days_employed'])
#Let's check that values of 'days_employed' now saved as positive:
print(credit_scoring[credit_scoring['days_employed'] < 0])
print(credit_scoring[credit_scoring['gender']=='XNA'])


Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []
Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []


In [5]:
#We can see that while checking 'children' there are entries with negative values.
print(credit_scoring['children'].value_counts())
print()
print((credit_scoring[credit_scoring['children'] < 0]).count())
print()
#There are 47 negative values that seems to be an error of data saving, so we can turn them to positive
pd.to_numeric(credit_scoring['children'], errors = 'coerce')
credit_scoring['children'] = abs(credit_scoring['children'])
#In addition, there are 76 entries that have value ‘20’ children that seems like another error and should be changed to '2'
credit_scoring['children'] = credit_scoring['children'].replace(20, 2)
print(credit_scoring['children'].value_counts())


 0     14148
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

children            47
days_employed       44
dob_years           47
education           47
education_id        47
family_status       47
family_status_id    47
gender              47
income_type         47
debt                47
total_income        44
purpose             47
dtype: int64

0    14148
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64


### Conclusion

While checking 'days_employed' it was found 15905 entries with negative values that seems to be an error of data saving, so they were updated to positive. Another set of 47 negative values was found in column 'children' and turned to positive. In addition, there were 76 entries that had '20' children. This was updated to '2'. 

### Processing duplicates

In [6]:
#Let's take a look on client's age 'dob_years' column. 
#print(credit_scoring['dob_years'].value_counts())
#We can see 101 entries with value '0', it'd be better to fill with average value, so we keep the data:
print(round(credit_scoring['dob_years'].mean()))
credit_scoring['dob_years'] = credit_scoring['dob_years'].replace(0, (round(credit_scoring['dob_years'].mean())))
#print(credit_scoring['dob_years'].value_counts())

43


In [7]:
#Let's take a look on client's 'education' column:
print(credit_scoring['education'].value_counts())
#There are values with different spelling and high/low case that means the same in the ‘education’ column:
credit_scoring['education'] = credit_scoring['education'].str.lower()
print(credit_scoring['education'].value_counts())

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             667
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64
secondary education    15233
bachelor's degree       5260
some college             743
primary education        282
graduate degree            6
Name: education, dtype: int64


In [8]:
#Let'simport library for stemming 
import nltk
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english')

#then split clients' 'purposes' into single words in the variable stemmed_word
queries = credit_scoring['purpose']
for query in queries:
    for word in query.split(' '):
        stemmed_word = english_stemmer.stem(word)
        #print('Source word -', word, ', after stemming -', stemmed_word)
        #print(stemmed_word)

#now we can see what are the most frequent ones in order to find out categories
        from nltk.stem import WordNetLemmatizer
        from collections import Counter
        wordnet_lemma = WordNetLemmatizer()
        words = nltk.word_tokenize(stemmed_word)
        lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
 
        from collections import Counter
        counter = Counter(lemmas)
        
        #print(list(set(dict.fromkeys(counter))))
        #print((sorted(counter, key=counter.get, reverse=True)))
#new table
        final = (sorted(counter, key=counter.get, reverse=True))
        #for element in final:
            #print(element) #(list(dict.fromkeys(element)))


In [9]:
# Imagine that import is in a first cell
# from nltk.stem import SnowballStemmer

english_stemmer = SnowballStemmer('english')

words = credit_scoring['purpose'].str.cat(sep=' ').split()
stemmed_words = []

for word in words:
    stemmed_words.append(english_stemmer.stem(word))
    
Counter(stemmed_words)

Counter({'purchas': 3314,
         'of': 2998,
         'the': 1288,
         'hous': 3820,
         'car': 4315,
         'supplementari': 909,
         'educ': 3526,
         'to': 3081,
         'have': 1551,
         'a': 5130,
         'wed': 2348,
         'transact': 2610,
         'for': 1294,
         'my': 2396,
         'famili': 641,
         'buy': 4003,
         'real': 4477,
         'estat': 4477,
         'commerci': 1315,
         'residenti': 607,
         'construct': 635,
         'own': 2240,
         'properti': 2542,
         'build': 1246,
         'second-hand': 968,
         'with': 1281,
         'becom': 412,
         'get': 1316,
         'an': 443,
         'ceremoni': 797,
         'higher': 426,
         'profil': 436,
         'univers': 949,
         'rent': 653,
         'out': 653,
         'renov': 612,
         'go': 496})

### Conclusion

While checking duplicates, we found values with different high/low case that means the same in the ‘education’ column and alligned them with str.lower() function. In addition, in the 'dob_years' column value '0' was replaced with the average age in order to improve data quality ('0' doesn't seem to be a valid age to get the credit). We could see values with different wording but same meaning in the purpose of credit column 'purpose', so it was grouped by replacing them as follows: car, wedding, education, real estate.

### Categorizing Data

In [10]:
#It might be interesting to classify data 
#by AGE GROUPS: YOUNG, ADULT, SENIOR
#The age group is returned according to the "dob_years" value, by using the following rules:
#'young' with age value <= 25 years
#'senior' with age value >= 64 years
#'adult' for all other cases 
def age_group(row):
    dob_years = row['dob_years']
    
    if dob_years <= 25:
        return 'young'
    elif dob_years >= 64:
        return 'senior'
    return 'adult'
credit_scoring['age_group'] = credit_scoring.apply(age_group, axis=1)
print(credit_scoring['age_group'].value_counts())
#There are 19127 adult clients, 1233 young and 1164 senior

adult     19127
young      1233
senior     1164
Name: age_group, dtype: int64


In [11]:
#by credit PURPOSE: WEDDING, CAR, REAL ESTATE, EDUCATION
print(credit_scoring['purpose'].value_counts())
#There are 10839 clients who took a credit for real estate, 4315 to buy a car, 4022 to get the education and 2348 for wedding

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
buying property for renting out             653
housing transactions                        653
transactions with commercial real estate    651
housing                                     647
purchase of the house                       647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             623
purchase of my own house                    620
building a property                         620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

In [12]:
#by whether the credit is paid on time - DEBT: YES, NO
print(credit_scoring['debt'].value_counts())
#There are 19783 clients who pay on time and 1741 who are not 

0    19783
1     1741
Name: debt, dtype: int64


In [13]:
#by FAMILY STATUS: SINGLE, NOT SINGLE (merge all clients who have a couple vs those who are alone to check whether there is a correlation and it's harder to repay debt when someone lives alone)
#First, let's check what id corresponds to what status:
credit_scoring.groupby(['family_status_id','family_status']).size()
#0 - married, 1 - civil partnership, 2 - widow / widower, 3 - divorced, 4 - unmarried
#Thus, there will be two groups: single - 2, 3, 4, not single - 0, 1
def family_group(row):
    family_status = row['family_status_id']
    
    if family_status <= 1:
        return 'not single'
    return 'sinlge'

credit_scoring['family_group'] = credit_scoring.apply(family_group, axis=1)
print(credit_scoring['family_group'].value_counts())  
#There are 16556 clients who live in couples an 4968 who are single

not single    16556
sinlge         4968
Name: family_group, dtype: int64


In [14]:
#by childfree or having CHILDREN: NO CHILDREN, CHILDREN (merge all clients who have kids and those who are childfree to check whether there is a correlation and it's harder to repay debt when there are kids)
#There will be two groups: childfree  - 0, children - all other values
def children_group(row):
    children = row['children']
    
    if ( children == 0):
        return 'childfree'
    return 'children'

credit_scoring['children_group'] = credit_scoring.apply(children_group, axis=1)
print(credit_scoring['children_group'].value_counts())  
#There are 14148 clients who have no children and 7376 who have at list one kid 

childfree    14148
children      7376
Name: children_group, dtype: int64


In [15]:
#by education: , CHILDREN (merge all clients who have kids and those who are childfree to check whether there is a correlation and it's harder to repay debt when there are kids)
#First, let's check what id corresponds to what status:
credit_scoring.groupby(['education_id','education']).size()

# 0 - bachelor's degree, 1 - secondary education, 2 - some college, 3 - primary education, 4 - graduate degree
#Thus, there will be two groups: graduated  - 0, 4, undergraduated - 1, 2, 3
def education_group(row):
    education = row['education_id']
    
    if ( 0 < education < 4):
        return 'undergraduated'
    return 'graduated'

credit_scoring['education_group'] = credit_scoring.apply(education_group, axis=1)
print(credit_scoring['education_group'].value_counts())  
#There are 16258 clients who did not finish the university and 5266 who did 

#print(credit_scoring.head(10))
#print(credit_scoring[credit_scoring['family_group'] == 'sinlge'].head())

undergraduated    16258
graduated          5266
Name: education_group, dtype: int64


### Conclusion

While our credit data was classified according to the age groups, education level, whether people are living in couples or having kids, we found that there are :
- 19127 adult clients, 1233 young and 1164 senior
- 19783 clients who pay on time and 1741 who are not 
- 10839 clients who took a credit for real estate, 4315 to buy a car, 4022 to get the education and 2348 for wedding
- 14148 clients who have no children and 7376 who have at list one kid
- 16556 clients who live in couples and 4968 who are single
- 16258 clients who did not finish the university and 5266 who did 

In the next part let's take a more precise look on our clients comparing number of those characteristics between each other.

### Step 3. -1. Is there a relation between having kids and repaying a loan on time?

In [16]:

#In order to create a pivote table, let's create a new variable and a column 'debt_id' 
#with string 'yes/no' values 
def debt_id(row):
    debt = row['debt']
    if debt == 1 :  
        return 'yes'
    return 'no'
credit_scoring['debt_id'] = credit_scoring.apply(debt_id, axis=1)

#display(credit_scoring.head())
data_pivot = credit_scoring.pivot_table(index=['children_group'], columns='debt_id', values='debt', aggfunc='count')
data_pivot['total'] = data_pivot['yes'] + data_pivot['no']
data_pivot['ratio_debtVStotal,%'] = round(data_pivot['yes'] / data_pivot['total']*100,1)

print(data_pivot)
#Those clients who have no kids don't repay their loan in 7.5% of cases, vs. 9.2% of those who have kids.

debt_id            no   yes  total  ratio_debtVStotal,%
children_group                                         
childfree       13085  1063  14148                  7.5
children         6698   678   7376                  9.2


### Conclusion

When we go deeper to our analysis, we can see that there are 20846 cases when having kids doesn't impact the loan repaiment and 678 cases when there is an impact, that represent 3% of cases.
Those clients who have no kids don't repay loan their in 7.5% of cases, vs. 9.2% of those who have kids.

### Step 3. -2. Is there a relation between marital status and repaying a loan on time?

In [17]:

#-----------------------
data_pivot = credit_scoring.pivot_table(index=['debt_id'], columns='family_group', values='debt', aggfunc='count')
data_pivot['total_group'] = data_pivot['not single'] + data_pivot['sinlge']
data_pivot['ratio_singleVStotal_group,%'] = round(data_pivot['sinlge'] / data_pivot['total_group']*100,1)

print(data_pivot)
#Among those clients who have no debt, 23% are single (widow / widower, divorced or unmarried)
#whereas among those clients who have delay in their loan repayment, 24% are single.

family_group  not single  sinlge  total_group  ratio_singleVStotal_group,%
debt_id                                                                   
no                 15237    4546        19783                         23.0
yes                 1319     422         1741                         24.2


In [18]:
# before revision 1/7/2020:
#-----------------------
#In order to check how the marital status and repaying loan on time are correlated, let's create a new variable marital_loan_trust
#So we have to find those who have debt ('debt'==1) and have marital status - married or in civil partnership (family_group == 'not single') vs those who carry this loan alone
def marital_loan_trust(row):
    family_group = row['family_group']
    debt = row['debt']  
    if family_group == 'sinlge' :
        if debt == 1 :
            return 'single_debt_correlated'
    if family_group == 'not single' :
        if debt == 1 :
            return 'couple_debt_correlated'
    
    return 'marital_debt_not_correlated'

#Then let's add this as column marital_loan_trust to credit_scoring table and calculate the percent of total number of cases
credit_scoring['marital_loan_trust'] = credit_scoring.apply(marital_loan_trust, axis=1)
#print(credit_scoring[credit_scoring['marital_loan_trust'] == 'single_debt_correlated'].head())
couple_debt_correlated = (credit_scoring[credit_scoring['marital_loan_trust'] == 'couple_debt_correlated']).count()/len(credit_scoring['debt'])
single_debt_correlated = (credit_scoring[credit_scoring['marital_loan_trust'] == 'single_debt_correlated']).count()/len(credit_scoring['debt'])

print('Out of total number of clients, those who have a delay in repaying a loan living:')
print('- in couples (married or in civil partnership): {:.1%}'.format(couple_debt_correlated['marital_loan_trust']))
print('- alone (widow / widower, divorced or unmarried): {:.1%}'.format(single_debt_correlated['marital_loan_trust']))


#There are 20846 cases when having kids doesn't impact the loan repaiment and 678 cases when there is an impact.
#To double check our result, let see the table .size() method:
#credit_scoring.groupby(['family_group','debt']).size()

Out of total number of clients, those who have a delay in repaying a loan living:
- in couples (married or in civil partnership): 6.1%
- alone (widow / widower, divorced or unmarried): 2.0%


### Conclusion

We can see that the marital situation has some correlation with the loan repaiment. There are 2% of th Out of total number of clients, those who have a delay in repaying a loan living:
- in couples (married or in civil partnership): 6.1%
- alone (widow / widower, divorced or unmarried): 2.0%

Among those clients who have no debt, 23% are single (widow / widower, divorced or unmarried), whereas among those clients who have delay in their loan repayment, 24% are single.

### Step 3. -3. Is there a relation between income level and repaying a loan on time?

In [19]:
#To see what is the relation between income level and repaying a loan on time let's devide it on groups
print(credit_scoring['total_income'].value_counts())
#There are 19347 clients who reported their income with there are 21525 clients as we checked earlier

17312.717    2
31791.384    2
42413.096    2
54857.666    1
14929.232    1
            ..
48796.341    1
34774.610    1
15710.698    1
19232.334    1
9591.824     1
Name: total_income, Length: 19347, dtype: int64


In [20]:
#Might be interesting to check if there is a correlation between purpose of credit, then age group and not reported income
credit_scoring_income_isnull = credit_scoring[credit_scoring['total_income'].isnull()]
print(credit_scoring_income_isnull['purpose'].value_counts())
print(credit_scoring_income_isnull['age_group'].value_counts())
#it looks like all categories were allowed somehow to take a loan without reporting their income.

having a wedding                            92
to have a wedding                           81
wedding ceremony                            76
construction of own property                75
housing transactions                        74
buy real estate                             72
transactions with my real estate            71
purchase of the house for my family         71
housing renovation                          70
transactions with commercial real estate    70
buy commercial real estate                  67
buying property for renting out             65
property                                    62
buy residential real estate                 61
real estate transactions                    61
housing                                     60
building a property                         59
cars                                        57
going to university                         56
to become educated                          55
second-hand car purchase                    54
buying my own

In [21]:
#let's we'll examine those clients who did reported their income that is 90%:
credit_scoring_income = credit_scoring[credit_scoring['total_income'].notnull()]
print('Here we examine those clients who did reported their income that is {:.0%}'.format(len(credit_scoring_income)/len(credit_scoring)),'of total clients.')
#print(credit_scoring_income)

print('Maximum income is', round(credit_scoring_income['total_income'].max()))
print('Minimum income is', round(credit_scoring_income['total_income'].min()))
print('Average income is', round(credit_scoring_income['total_income'].mean()))
print('Median income is', round(credit_scoring_income['total_income'].median()))
print()

Here we examine those clients who did reported their income that is 90% of total clients.
Maximum income is 362497
Minimum income is 3307
Average income is 26787
Median income is 23202



In [22]:
#As average income is pretty close to the median income, 
#that means that our clients group income is spread pretty homogenous, 
#without significant gaps

#let's devide customers in 2 groups - below median and above median
#for this purpose let's create new variable income_median
income_median = credit_scoring_income['total_income'].median()
       
#and then define two groups: 
#income_below_median and income_above_median and then 
#save it in additional column of credit_scoring table
def income_group(row):
    total_income = row['total_income']
    debt = row['debt']
    
    if total_income <= income_median :
        if debt == 1 :
            return 'income_below_median_debt'
    if total_income <= income_median :
        if debt == 0 :
            return 'income_below_median_nodebt'
    if total_income > income_median :
        if debt == 0 :
            return 'income_above_median_nodebt'
    if total_income > income_median :
        if debt == 1 :
            return 'income_above_median_debt'

credit_scoring['income_group'] = credit_scoring.apply(income_group, axis=1)
print(credit_scoring['income_group'].value_counts())  

income_above_median_nodebt    8908
income_below_median_nodebt    8871
income_below_median_debt       804
income_above_median_debt       767
Name: income_group, dtype: int64


In [23]:
#let's see what is the relation among those who pay loan on time and those who don't:
#print('Clients who can not pay their loan on time represent {:.0%}'.format(len(credit_scoring[credit_scoring[['income_group'] == 'income_below_median_debt']]))/len(credit_scoring)),'of total clients.')
income_below_median_debt = (credit_scoring[credit_scoring['income_group'] == 'income_below_median_debt']).count()/len(credit_scoring_income['debt'])
income_above_median_debt = (credit_scoring[credit_scoring['income_group'] == 'income_above_median_debt']).count()/len(credit_scoring_income['debt'])

print('Out of total number of clients, who reported their income, those who have a delay in repaying a loan earning:')
print('- below median income: {:.0%}'.format(income_below_median_debt['income_group']))
print('- above median income: {:.0%}'.format(income_above_median_debt['income_group']))

#credit_scoring.groupby(['income_group','debt']).size()

Out of total number of clients, who reported their income, those who have a delay in repaying a loan earning:
- below median income: 4%
- above median income: 4%


### Conclusion

Here we examine those clients who did reported their income that is 90% of total clients.
Out these clients, those who have a delay in repaying a loan earning:
- below median income: 4.2%
- above median income:: 4.0%

### Step 3. -4. How do different loan purposes affect on-time repayment of the loan?

In [24]:
credit_scoring.groupby(['purpose','debt']).size()

#In order to take a look on on-time repayment of the loan 
#let's use variable ontime_debt_total that sums up those who has no debt ('debt' == 0)
ontime_debt_total = credit_scoring[credit_scoring['debt'] == 0]
print('Different loan purposes affect on-time repayment in ', (len(ontime_debt_total)),'cases as follows:')
print()
#Let's take a look for what purposes loans were repaid successfully
purpose = ontime_debt_total['purpose'].value_counts()/len(ontime_debt_total)

print('Real Estate  | Car     | Education | Wedding  ')
print('-----------------------------------------------')
print('{: >12.1%} | {: >7.1%} | {: >9.1%} | {: >9.1%} '.format(purpose[0], purpose[1], purpose[2], purpose[3]))
print()
print('In general, loan is repaid in {:.0%}:'.format(len(ontime_debt_total)/len(credit_scoring)),'cases')
print()

#credit_scoring.info()

Different loan purposes affect on-time repayment in  19783 cases as follows:

Real Estate  | Car     | Education | Wedding  
-----------------------------------------------
        3.7% |    3.6% |      3.6% |      3.1% 

In general, loan is repaid in 92%: cases



### Conclusion

Different loan purposes affect on-time repayment in 19783 cases as follows:
- real estate: 50.8% 
- car: 19.8%
- education: 18.5%
- wedding: 10.9% 

### Step 4. General conclusion

There is a data set credit_scoring.csv that was examined in this project. It consists of 21525 rows, we suppose that every row represents a client who took a loan. Among them we had to find those groups who repays on time depending on their marital status, having or not having kids, income level and  loan purposes.

Out of total number of clients, those who have a delay in repaying a loan living:
- in couples (married or in civil partnership): 6.1%
- alone (widow / widower, divorced or unmarried): 2.0%

There is a relation between having kids and delay in repaying a loan on time in 3.1% cases.

We examined those clients who did reported their income that is 90% of total clients. Out these clients, those who have a delay in repaying a loan earning:
* below median income: 4.2%
* above median income: 4.0%

As per those, who repay there loan on time, loan purposes are divided as follows::
* real estate: 50.8% 
* car: 19.8%
* education: 18.5%
* wedding: 10.9% 

In general, loan is repaid in 92% of cases.


Below you can find more details regarding data preprocessing.
Predefined columns: 

The clients’ family status appears in the ‘family_status’ column and 'family_status_id'; their total income, in the ‘total_income’ column (as float); and their ages (as integers), in the ‘dob_years’ column; number of children in 'children' column, employment period in 'days_employed', educational level in 'education' column and 'education_id', client's gender in 'gender' column; source of income in 'income_type'; status of the credit repayment in 'debt' column; credit's purpose in 'purpose'.


For analysis purposes following columns were created in order to deep dive into groups:
- age_group 
- family_group
- children_group   
- education_group 
- kids_loan_trust 
- marital_loan_trust
- income_group.

While cleaning the data following manipulations were executed:
- missing values were replaced in case of ‘gender’ column - 1 row, value XNA
- erroneous data was fixed in ‘children’ column:  
    - negative value replaced with positive
    - value ’20’ replaced by ‘2’ - as it had significant number of clients and pretty unusual, taking into account that the next number of kids was ‘5’; thus it was considered as ‘2’ that was saved with errors 
- negative values in ‘days_employed’ column were replace with positive
- data in column ’education’ was aligned with lower case
- column ‘purpose’ were aligned according to 4 categories: car, wedding, real estate, education; all other variations were replaced by those four
- column ‘tatal_income’ had 10% of missing values, thus this group (without missing values) was examined separately in the dedicated part regarding relation between loan repayment and income. 
